tools/luthier.luau

```diff
@@ -0,0 +1,437 @@
+	if success then
+		return res
+	else
+		return nil
+	end
+end
+
+local function getSourceRoot()
+	local cwd = process.cwd()
+
+	while true do
+		if safeFsType(cwd .. pathSep .. ".LUTE_SENTINEL") == "file" then
+			return cwd
+		end
+
+		local lastSep = findLast(cwd, pathSep)
+		if lastSep then
+			cwd = string.sub(cwd, 1, lastSep - 1)
+		else
+			break
+		end
+	end
+
+	if process.env.LUTE_ROOT_DIR then
+		return process.env.LUTE_ROOT_DIR
+	end
+
+	error(
+		"Could not locate lute project root.  Your cwd must be in your lute checkout, or the LUTE_ROOT_DIR environment variable must be set."
+	)
+end
+
+local function getExeName(target: string): string
+	local baseName = targetMap[target].exeName
+	local ext = if isWindows then ".exe" else ""
+	return baseName .. ext
+end
+
+local function getCompiler(): string
+	if isMac then
+		return "xcode"
+	else
+		return "vs2022"
+	end
+end
+
+local function getConfig(): string
+	return args:get("config")
+end
+
+local function getBuildDir(): string
+	local config = string.lower(getConfig())
+
+	if isLinux then
+		return joinPath(BUILD_DIR, config)
+	else
+		return joinPath(BUILD_DIR, getCompiler(), config)
+	end
+end
+
+local function getProjectPath(): string
+	return projectRelative(getBuildDir())
+end
+
+local function projectPathExists(): boolean
+	local projectPath = getProjectPath()
+
+	return safeFsType(projectPath) == "dir"
+end
+
+local function getStdLibHash(): string
+	local libsPath = projectRelative("std", "libs")
+	local libs = fs.listdir(libsPath)
+
+	local toHash = ""
+	for _, lib in libs do
+		toHash ..= lib.name
+		toHash ..= fs.readfiletostring(joinPath(libsPath, lib.name))
+	end
+
+	return toHash
+end
+
+local function isGeneratedStdLibUpToDate(): boolean
+	local hashFile = projectRelative("std", "src", "generated", "hash.txt")
+
+	if safeFsType(hashFile) ~= "file" then
+		return false
+	end
+
+	local hash = fs.readfiletostring(hashFile)
+
+	return hash == getStdLibHash()
+end
+
+local function writeLines(file, lines: { string })
+	fs.write(file, table.concat(lines, "\n") .. "\n")
+end
+
+local function generateStdLibFilesIfNeeded()
+	local generatedPath = projectRelative("std", "src", "generated")
+	pcall(fs.mkdir, generatedPath)
+
+	if isGeneratedStdLibUpToDate() then
+		return
+	end
+
+	print("Generating code for @std libraries, files are out of date.")
+
+	local cpp = fs.open(joinPath(generatedPath, "modules.cpp"), "w+")
+
+	writeLines(cpp, {
+		"// This file is auto-generated by luthier.luau. Do not edit.",
+		"// Instead, you should modify the source files in `std/libs`.",
+		"",
+		'#include "modules.h"',
+		"",
+		"constexpr std::pair<std::string_view, std::string_view> lutestdlib[] = {",
+	})
+
+	local libsPath = projectRelative("std", "libs")
+	local libs = fs.listdir(libsPath)
+
+	for _, lib in libs do
+		local aliasedPath = `@std/{lib.name}`
+		local libSrc = fs.readfiletostring(joinPath(libsPath, lib.name))
+		fs.write(cpp, `\n    \{"{aliasedPath}", R"({libSrc})"},\n`)
+	end
```

Comment:
Doesn't support directories in std atm. No equivalent to os.walk makes this annoying, and it's unnecessary at the moment.